### ISSUE TITLE
Something's barfing when I try ido-kill-buffer-at-head

### ISSUE THREAD

**Issue by thwh** [minad/vertico#361]
First of all: these packages are great. Thank you. I think they'll eventually be standard emacs.

TLDR: I miss `ido-kill-buffer-at-head` in consult-buffer and vertico-flavored `ido-switch-buffer`.

I prefer to use consult-buffer, but I found there was no equivalent of `ido-kill-buffer-at-head`: I know I can run embark, and then kill it, but it's another step and it additionally ends the -switch-buffer session, meaning I can't just quickly clean up the buffer list. Also I had to tell embark I don't consider it a 'dangerous' command, like this:

```
(add-to-list 'embark-pre-action-hooks '(kill-buffer . nil))
```

All of which seems like too much trouble for a relatively simple action.

So, I've tried using ido-switch-buffer with vertico, and adding back the command like this:

```
(keymap-set vertico-map "C-k" #'ido-kill-buffer-at-head)
````

But something's sending a string to that method I guess. I get:

`ido-kill-buffer-at-head: Wrong type argument: stringp, nil`

I just tried enabling vertico and consult progressively with emacs -Q, and couldn't find the culprit.

Here's my setup, with all relevant packages:

```
(use-package vertico
  :init
  (vertico-mode)

  (setq vertico-cycle t)

  :config
  (keymap-set vertico-map "C-k" #'ido-kill-buffer-at-head)

  ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
  ;; Extensions:
  ;; https://github.com/minad/vertico#extensions
  (keymap-set vertico-map "M-c" #'vertico-quick-exit)
  (keymap-set vertico-map "RET" #'vertico-directory-enter)
  (keymap-set vertico-map "DEL" #'vertico-directory-delete-char)
  (keymap-set vertico-map "M-DEL" #'vertico-directory-delete-word)
  (keymap-set vertico-map "M-i" #'vertico-directory-delete-word)
  (add-hook 'rfn-eshadow-update-overlay-hook #'vertico-directory-tidy)

  ;; https://github.com/minad/vertico/#configure-vertico-per-command-or-completion-category
  (vertico-multiform-mode)

  ;; Configure the display per command.
  (setq vertico-multiform-commands
        '(
          ;; 2 M-x versions:
          (execute-extended-command flat)
          (consult-execute-extended-command-list indexed buffer)

          ;; 2 buffer versions:
          ;; (ido-switch-buffer flat)
          (consult-buffer flat (vertico-cycle . t))
          (consult-buffer-list indexed buffer)

          (consult-imenu buffer indexed)
          (consult-grep buffer)
          ))

  ;; ;; Configure the display per completion category.
  (setq vertico-multiform-categories
        '(
          (file indexed)
          ))
)

(use-package consult
  ;; Replace bindings. Lazily loaded due by `use-package'.
  :bind (
         ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
         ;; Trying not to bind anything I won't use:
         ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
         ;; C-c bindings in `mode-specific-map'
         ("M-X" . consult-mode-command)
         ([remap Info-search] . consult-info)

         ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
         ;; C-x bindings in `ctl-x-map'
         ("M-~" . consult-buffer)
         ("C-x b" . consult-buffer-list)
         ("C-x M-:" . consult-complex-command)

         ;; Fast register access
         ("M-#" . consult-register-load)
         ("M-'" . consult-register-store)

         ;; Other custom bindings
         ("M-y" . consult-yank-pop)

         ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
         ;; M-g bindings in `goto-map'
         ("M-g m" . consult-mark)
         ("M-g M" . consult-global-mark)

         ("M-g e" . consult-compile-error)
         ("M-g g" . consult-goto-line)
         ("M-g i" . consult-imenu)

         ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
         ;; override isearch
         ("C-s" . consult-line)

         ;; M-s bindings in `search-map'
         ("M-s l" . consult-line)
         ("M-s ." . consult-line-thing-at-point)
         ("M-s g" . consult-grep)
         ("M-s k" . consult-keep-lines)
         ("M-s f" . consult-focus-lines)
         ("M-s d" . consult-find)

         ;; Isearch integration
         ("M-s e" . consult-isearch-history)

         :map isearch-mode-map
         ("M-s e" . consult-isearch-history)       ;; orig. isearch-edit-string
         ("M-s l" . consult-line)                  ;; needed by consult-line to detect isearch
         ("M-s L" . consult-line-multi)            ;; needed by consult-line to detect isearch

         ;; Minibuffer history
         :map minibuffer-local-map
         ("M-s" . consult-history)                 ;; orig. next-matching-history-element
         ("M-r" . consult-history)                ;; orig. previous-matching-history-element
         )

  :hook
  ;; Enable automatic preview at point in the *Completions* buffer. This is
  ;; relevant when you use the default completion UI.
  (completion-list-mode . consult-preview-at-point-mode)

  :init

  ;; Optionally configure the register formatting. This improves the register
  ;; preview for `consult-register', `consult-register-load',
  ;; `consult-register-store' and the Emacs built-ins.
  (setq register-preview-delay 0.5
        register-preview-function #'consult-register-format)

  ;; Optionally tweak the register preview window.
  ;; This adds thin lines, sorting and hides the mode line of the window.
  (advice-add #'register-preview :override #'consult-register-window)

  ;; Use Consult to select xref locations with preview
  (setq xref-show-xrefs-function #'consult-xref
        xref-show-definitions-function #'consult-xref)

  :config

  ;; https://github.com/minad/consult#fine-tuning-of-individual-commands

  ;; For some commands and buffer sources it is useful to configure the
  ;; :preview-key on a per-command basis using the `consult-customize' macro.

  ;; The default value is 'any, such that any key triggers the preview.
  ;; (setq consult-preview-key 'any)
  ;; or make it manual:
  ;; (setq consult-preview-key "M-.")

  (consult-customize
   consult-line
   :add-history (seq-some #'thing-at-point '(region symbol))
   ;; :preview-key "M-."
   :preview-key 'any
   )

  ;; intended as a replacement for M-s .
  (defalias 'consult-line-thing-at-point 'consult-line)
  (consult-customize
   consult-line-thing-at-point
   :initial (thing-at-point 'symbol)
   :preview-key 'any
   )

  ;; alternative to C-b flat, with fast preview
  (defalias 'consult-buffer-list 'consult-buffer)
  (consult-customize
   consult-buffer-list
   :preview-key 'any
   )

  (consult-customize
   consult-buffer
   :preview-key "M-."
   )

  ;; fast but not immediate:
  (consult-customize
   consult-theme :preview-key '(:debounce 0.2 any)
   consult-ripgrep consult-git-grep
   consult-bookmark consult-recent-file
   consult--source-bookmark
   consult--source-recent-file consult--source-project-recent-file
   :preview-key '(:debounce 0.4 any)
   )

  ;; < ? : show options
  (setq consult-narrow-key "<")

  ;; Optionally make narrowing help available in the minibuffer.
  ;; You may want to use `embark-prefix-help-command' or which-key instead.
  (define-key consult-narrow-map (vconcat consult-narrow-key "?") #'consult-narrow-help)

)


(use-package orderless
  :ensure t
  :custom
  (completion-styles '(orderless basic))
  (completion-category-overrides '((file (styles basic partial-completion)))))


(use-package marginalia
  :ensure t
  :config
  (marginalia-mode))

(use-package embark
  :ensure t

  :bind
  (
   ;; Bound to C-. via karabiner:
   ("M-*" . embark-act)
   ;; should mostly behave like M-. where we expect it to:
   ("M-." . embark-dwim)
   ("C-h B" . embark-bindings)  ;; alternative for `describe-bindings'
   )

  :init

  ;; Optionally replace the key help with a completing-read interface
  (setq prefix-help-command #'embark-prefix-help-command)

  :config

  ;; don't make me confirm killing a buffer:
  (add-to-list 'embark-pre-action-hooks '(kill-buffer . nil))

  ;; Hide the mode line of the Embark live/completions buffers
  (add-to-list 'display-buffer-alist
               '("\\`\\*Embark Collect \\(Live\\|Completions\\)\\*"
                 nil
                 (window-parameters (mode-line-format . none))))

)

```

**Comment by thwh** [comment]
Wow that was fast. It looks like `embark-quit-after-action` can be set per command, as an alist? So I'd make my own little macro calling `embark-act` > `kill-buffer`, bind it to `C-k` or whatever, and then set up `embark-quit-after-action` to look something like:

```
((embark-kill-buffer-at-head nil) t)
```

**Comment by minad** [comment]
Yes, except that you write `kill-buffer` instead of `embark-kill-buffer-at-head`.

**Comment by thwh** [comment]
Because we're talking about what embark actually sees, not my macro, right...

**Comment by thwh** [comment]
But: I still don't quite understand why `ido-kill-buffer-at-head` isn't working when I'm running `ido-switch-buffer` : I thought vertico was supposed to work with the defaults? Or is ido not considered vanilla emacs, really?

**Comment by thwh** [comment]
Oh god you're right, it's right there:

https://github.com/oantolin/embark#quitting-the-minibuffer-after-an-action

In any case, thank you for the prompt response. I'm sure I'll figure it out. And thank you again for your elisp.

**Comment by minad** [comment]
> But: I still don't quite understand why ido-kill-buffer-at-head isn't working when I'm running ido-switch-buffer : I thought vertico was supposed to work with the defaults? Or is ido not considered vanilla emacs, really?

Ido can be considered obsolete. It doesn't implement `completing-read` fully.

**Comment by thwh** [comment]
Understood.

Got it working. For anyone else should they find this:

```
  ;; don't quit:
  (setq embark-quit-after-action '((kill-buffer . nil) (t . t)))

  ;; don't make me confirm:
  (add-to-list 'embark-pre-action-hooks '(kill-buffer . nil))

  (defalias 'embark-kill-buffer-at-head
    (kmacro "M-* k"))

  ;; add to vertico-map:
  (keymap-set vertico-map "C-k" #'embark-kill-buffer-at-head)
```

**Comment by minad** [comment]
Isn't it possible to directly write this?

~~~ elisp
(keymap-set vertico-map "C-k" "M-* k")
~~~
